src/authz-smart/smartHandler.ts

```diff
@@ -8,31 +8,130 @@ import {
     AllowedResourceTypesForOperationRequest,
+    SystemOperation,
+    TypeOperation,
 } from 'fhir-works-on-aws-interface';
-import { SMARTConfig } from './smartConfig';
+import axios from 'axios';
+import { LaunchType, ScopeType, SMARTConfig } from './smartConfig';
 
 // eslint-disable-next-line import/prefer-default-export
 export class SMARTHandler implements Authorization {
+    static readonly SCOPE_REGEX = /(patient|user|system|launch)\/(\w+|\*)\.?(\w+|\*)?|(^launch$)/;
+
     private readonly version: number = 1.0;
 
     private readonly config: SMARTConfig;
 
     constructor(config: SMARTConfig) {
-        this.config = config;
-        if (this.config.version !== this.version) {
-            throw Error('Configuration version does not match handler version');
+        if (config.version !== this.version) {
+            throw Error('Authorization configuration version does not match handler version');
         }
+        this.config = config;
     }
 
     async isAuthorized(request: AuthorizationRequest): Promise<boolean> {
+        const decoded = decode(request.accessToken, { json: true }) || {};
+        const { aud, iss } = decoded;
+        // verify aud & iss
+        if (this.config.expectedAudValue !== aud || this.config.expectedIssValue !== iss) {
+            console.error('aud or iss is not matching');
+            return false;
+        }
+        // verify scope
+        let scopes: string[] = [];
+        if (this.config.scopeValueType === 'space' && typeof decoded[this.config.scopeKey] === 'string') {
+            scopes = decoded[this.config.scopeKey].split(' ');
+        } else if (this.config.scopeValueType === 'array' && Array.isArray(decoded[this.config.scopeKey])) {
+            scopes = decoded[this.config.scopeKey];
+        }
+        if (!this.isScopeSufficient(scopes, request.operation, request.resourceType)) {
+            console.error('scopes are insuffiecient');
+            console.error(`scopes: ${scopes}`);
+            console.error(`operation: ${request.operation}`);
+            console.error(`resourceType: ${request.resourceType}`);
+            return false;
+        }
+
+        // Verify token
+        if (this.config.authZUserInfoUrl) {
+            console.log('Posting to AuthZ for more customer information');
+            let result;
+            try {
+                result = await axios.post(
+                    this.config.authZUserInfoUrl,
+                    {},
+                    { headers: { Authorization: `Bearer ${request.accessToken}` } },
+                );
+            } catch (e) {
+                console.error('Post to authZUserInfoUrl failed');
+                console.error(e);
+            }
+            if (!result || !result.data[this.config.expectedFhirUserClaimKey]) {
+                console.error(`result from AuthZ did not have ${this.config.expectedFhirUserClaimKey} claim`);
+                return false;
+            }
+        }
         return true;
     }

```

Comment:
Are we ok with making a call to /userInfo on every request? This is expensive

Isn't it possible to have the FhirUser claim in the access token to begin with?

Also, this code has a weird order in that it analyzes claims/scopes of a possibly invalid token and then the token is validated w /userInfo at the end.

Reply:
All the JWT in the unit tests do have the `fhirUser` claim in them.

Reply:
Their are two ways to validate JWTs 
1. Decoding the JWT and verifying signature
    * We will need to call the AuthZ server for the public key - we could then cache that public key (outside of lambda)
    * This leaves the vulnerability of AuthZ revoking the accessToken and the FHIR API still accepting it as valid
1. Using Token introspection or userInfo endpoint
    * Downside here is network call for each AuthZ
    * We could move this to a custom authorizer (where we could cache) but same issue as above there are race conditions that would allow revoked accessTokens to still gain access
    * As a note the userInfo endpoint validates AND return user's claims

Yeah you are right you can add the fhirUser to the JWT access_token the trouble is you still need to verify it

As for code ordering, my thought is do the local checks first then do the expensive call -- I think ideally we use promises: kick off the call do local checks then check the response

https://developer.okta.com/docs/guides/validate-access-tokens/nodejs/overview/

Reply:
Why are we using `/userinfo` instead of `/introspect`?
`/introspect` is meant to validate tokens. 
`/userinfo` Is meant to provide user info and validates tokens as a side effect. Additional permissions scopes are required to call it and you are getting user info that the fhir server doesn't need (i.e. `family_name`)

In general designing for token revocation is a risk/performance balance. longer cache time and longer lived tokens give better performance but increase the risk of letting a revoked token go through.
Is the goal here to have zero tolerance for revoked tokens(and consequently the worst possible performance)?

Is this a generic implementation? Or is this Okta specific? Not all idps can revoke tokens to begin with.

Reply:
> Why are we using /userinfo instead of /introspect?

Some customer may not want to add `fhirUser` to the `access_token`, as it does leak **some** customer information. The alternative to adding it to the `access_token` is to use the `/userinfo` endpoint to ensure we get the `fhirUser` claim. Without this claim we do not know who they are.

> Is the goal here to have zero tolerance for revoked tokens(and consequently the worst possible performance)?

The goal with this PR is to solve the problem functionally first then optimize if there is customer interest/demand. An easy path of optimization is adding this logic into a Custom Authorizer and let API GW cache this information per JWT token.

> Is this a generic implementation? Or is this Okta specific? Not all idps can revoke tokens to begin with.

This should be a generic OAuth2/OIDC implementation. If you see Okta specific concerns please let me know!

Reply:
discussed offline a bit more. I'm ok with the `/userinfo` call here